Application/ControlMsrE2/ControlMsrE2Hii.c

```diff
@@ -0,0 +1,256 @@
+/** @file
+ Control MSR 0xE2 Hii HiiDatabase related stuff
+ 
+ Copyright (c) 2020, Brumbaer. All rights reserved.<BR>
+ This program and the accompanying materials
+ are licensed and made available under the terms and conditions of the BSD License
+ which accompanies this distribution.  The full text of the license may be found at
+ http://opensource.org/licenses/bsd-license.php
+ 
+ THE PROGRAM IS DISTRIBUTED UNDER THE BSD LICENSE ON AN "AS IS" BASIS,
+ WITHOUT WARRANTIES OR REPRESENTATIONS OF ANY KIND, EITHER EXPRESS OR IMPLIED.
+ 
+ **/
+
+#include "ControlMsrE2.h"
+
+typedef struct {
+    EFI_VARSTORE_ID     id;
+    EFI_IFR_VARSTORE   *varstoreHeader;
+} VarStoreContext;
+
+typedef VOID EFIAPI OpCodeHandler (EFI_IFR_OP_HEADER* IfrHeader, UINT8* Stop, VOID* Context);
+
+EFI_HII_PACKAGE_LIST_HEADER* HiiExportPackageLists (IN EFI_HII_HANDLE Handle) {
+    
+    UINTN BufferSize = 0;
+    
+    gHiiDatabase->ExportPackageLists (gHiiDatabase, Handle, &BufferSize, NULL);
+    
+    if (BufferSize > 0) {
+        EFI_HII_PACKAGE_LIST_HEADER* result = (EFI_HII_PACKAGE_LIST_HEADER*) AllocatePool (BufferSize);
+        
+        EFI_STATUS Status = gHiiDatabase->ExportPackageLists (gHiiDatabase, Handle, &BufferSize, result);
+        
+        return (Status) ? NULL : result;
+    }
+    return NULL;
+}
+
+EFI_IFR_OP_HEADER* DoForEachOpCode (
+                 IN EFI_IFR_OP_HEADER* Header,
+                 IN UINT8 OpCode,
+                 IN UINT8* Stop,
+                 IN void* Context,
+                 IN OpCodeHandler Handler
+                 ) {
+    
+    do {
+        if (Header->OpCode == EFI_IFR_END_OP)
+            return Header;
+        if (Header->OpCode == OpCode) {
+            Handler (Header, Stop, Context);
+            if ((Stop != NULL) && *Stop)
```

Comment:
If Stop can be NULL and is modified, maybe it should be `IN OUT UINT8 *Stop OPTIONAL` in the header?

Reply:
I'll have to check all the Specifiers